cmd/snap-confine/mount-support.c

```diff
@@ -524,29 +524,26 @@ static void sc_bootstrap_mount_namespace(const struct sc_mount_config *config)
 		sc_do_mount(SNAP_MOUNT_DIR, dst, NULL, MS_BIND | MS_REC, NULL);
 		sc_do_mount("none", dst, NULL, MS_REC | MS_SLAVE, NULL);
 	}
-	// Create the hostfs directory if one is missing. This directory is a part
-	// of packaging now so perhaps this code can be removed later.
-	sc_identity old = sc_set_effective_identity(sc_root_group_identity());
-	if (mkdir(SC_HOSTFS_DIR, 0755) < 0) {
-		if (errno != EEXIST) {
-			die("cannot perform operation: mkdir %s",
-			    SC_HOSTFS_DIR);
-		}
-	}
-	(void)sc_set_effective_identity(old);
-	// Ensure that hostfs isgroup owned by root. We may have (now or earlier)
-	// created the directory as the user who first ran a snap on a given
-	// system and the group identity of that user is visilbe on disk.
+	// Ensure that hostfs exists and is group-owned by root. We may have (now
+	// or earlier) created the directory as the user who first ran a snap on a
+	// given system and the group identity of that user is visible on disk.
 	// This was LP:#1665004
 	struct stat sb;
 	if (stat(SC_HOSTFS_DIR, &sb) < 0) {
-		die("cannot stat %s", SC_HOSTFS_DIR);
-	}
-	if (sb.st_uid != 0 || sb.st_gid != 0) {
-		if (chown(SC_HOSTFS_DIR, 0, 0) < 0) {
-			die("cannot change user/group owner of %s to root",
-			    SC_HOSTFS_DIR);
+		if (errno == ENOENT) {
+			// Create the hostfs directory if one is missing. This directory is a part
+			// of packaging now so perhaps this code can be removed later.
+			if (mkdir(SC_HOSTFS_DIR, 0755) < 0) {
+				die("cannot perform operation: mkdir %s", SC_HOSTFS_DIR);
+			}
+			if (chown(SC_HOSTFS_DIR, 0, 0) < 0) {
```

Comment:
If this directory could be created by a user on first-run, does this mean that the user (or the user's group?) would have write and execute access to the containing directory? In that case, the `mkdir` and `chown` could be raced to make a directory elsewhere, or change the permissions of another file, via a symlink created at the right moment.

Reply:
`SC_HOSTFS_DIR` is hardcoded to be `/var/lib/snapd/hostfs`; the code here can create only the last component of the path, because we assume (and this is a valid assumption, already present in the old code) that `/var/lib/snapd` has been created by the installation of the snapd package, so it will be chown root and have 0755 mode.

Even if we create our `/var/lib/snapd/hostfs` directory as an ordinary user, it's inside a directory owned by root, so an ordinary user cannot delete it and replace it with a symbolic link. Using `fchownat()` with the `AT_SYMLINK_NOFOLLOW` would make this code more robust from the readability point of view, because it would make evident that such an attach cannot happen, but with the knowledge we have of the `/var/lib/snapd/` setup we know that such an attack is not possible even now. I can change this code to use `fchownat()` if you prefer, I was just minimizing the changes from the previous code.

All this assuming I didn't misunderstand your comment:-)

Reply:
Okay, thanks; minimized changes is definitely a good idea. I *do* prefer `fchownat()` :) but if the directory can't be written by a user in the first place, it's fine. Thanks

Reply:
So I still worry about the chance of a race-condition now between the mkdir and chown calls - since in the future snap-confine will be running as a standard user, then that same user could then write to this new hostfs directory before the chown() call happens - how about setting the umask() first before the mkdir() so that there is no write permission to the new hostfs directory - then after doing the chown() you can chmod() it as needed? That should avoid unprivileged users being able to mess with the directory inbetween.

Reply:
Thanks for the suggestion, updated!